File: testing/components/services/dmtools_cli_service.js
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const JOB_RUNNER_MAIN_CLASS = 'com.github.istin.dmtools.job.JobRunner';
const JOB_RUNNER_JAVA_ARGS = [
  '-Dlog4j2.configurationFile=classpath:log4j2-cli.xml',
  '-Dlog4j.configuration=log4j2-cli.xml',
  '-Dlog4j2.disable.jmx=true',
  '-Djava.net.preferIPv4Stack=true',
  '--add-opens',
  'java.base/java.lang=ALL-UNNAMED',
  '-XX:-PrintWarnings',
  '-Dpolyglot.engine.WarnInterpreterOnly=false',
];

class DmtoolsCliService {
  constructor({ repositoryRoot, processRunner, options = {} }) {
    this.repositoryRoot = repositoryRoot;
    this.processRunner = processRunner;
    this.dmtoolsScriptPath =
      options.dmtoolsScriptPath || path.join(repositoryRoot, 'dmtools.sh');
    this.timeoutMs = options.timeoutMs || 120000;
    this.javaCommand = options.javaCommand || 'java';
    this.homeDirectory =
      options.homeDirectory || fs.mkdtempSync(path.join(os.tmpdir(), 'dmtools-cli-home-'));
    this.jarFile = options.jarFile || this.resolveJarFile();
  }

  runDmtools(args, options = {}) {
    return this.processRunner.run(this.dmtoolsScriptPath, args, {
      cwd: options.cwd || this.repositoryRoot,
      encoding: 'utf8',
      timeoutMs: options.timeoutMs || this.timeoutMs,
      env: {
        ...process.env,
        HOME: this.homeDirectory,
        ...options.env,
      },
    });
  }

  runJob(jobName, params = {}, options = {}) {
    const payload = Buffer.from(
      JSON.stringify({
        name: jobName,
        params,
      }),
    ).toString('base64');

    return this.processRunner.run(
      this.javaCommand,
      [...JOB_RUNNER_JAVA_ARGS, '-cp', this.jarFile, JOB_RUNNER_MAIN_CLASS, payload],
      {
        cwd: options.cwd || this.repositoryRoot,
        encoding: 'utf8',
        timeoutMs: options.timeoutMs || this.timeoutMs,
        env: {
          ...process.env,
          HOME: this.homeDirectory,
          ...options.env,
        },
      },
    );
  }

  combinedOutput(result) {
    return [result.stdout, result.stderr].filter(Boolean).join('\n').trim();
  }

  resolveJarFile() {
    const candidateDirectories = [
      path.join(this.repositoryRoot, 'build', 'libs'),
      path.join(this.repositoryRoot, 'dmtools-core', 'build', 'libs'),
    ];

    for (const directory of candidateDirectories) {
      const jarFile = findFirstFatJar(directory);
      if (jarFile) {
        return jarFile;
      }
    }

    throw new Error(
      [
        `Could not find a checkout-built DMTools JAR for ${this.repositoryRoot}.`,
        `Searched: ${candidateDirectories.join(', ')}`,
        'Build the current checkout first with ./gradlew :dmtools-core:shadowJar.',
        'Global ~/.dmtools/dmtools.jar fallback is intentionally disabled for this test.',
      ].join(' '),
    );
  }
}

function findFirstFatJar(directory) {
  if (!directory || !fs.existsSync(directory) || !fs.statSync(directory).isDirectory()) {
    return null;
  }

  const jarFiles = fs
    .readdirSync(directory)
    .filter((fileName) => fileName.endsWith('-all.jar'))
    .sort();

  return jarFiles.length > 0 ? path.join(directory, jarFiles[0]) : null;
}

module.exports = { DmtoolsCliService };

File: testing/components/services/dmtools_cli_service.py
from __future__ import annotations

import json
import re
import tempfile
from pathlib import Path
from typing import Any

from testing.core.interfaces.dmtools_cli import DmtoolsCli
from testing.core.interfaces.process_runner import ProcessRunner
from testing.core.models.process_execution_result import ProcessExecutionResult


class DmtoolsCliService(DmtoolsCli):
    COMPATIBILITY_RESPONSE = (
        "CodeGenerator compatibility shim executed successfully. "
        "No action was taken and no code artifacts were produced."
    )
    _NETWORK_ACTIVITY_PATTERN = re.compile(r"(connect|sendto|sendmsg)\(")

    def __init__(
        self,
        repository_root: Path,
        runner: ProcessRunner,
    ) -> None:
        self.repository_root = repository_root
        self.runner = runner
        self.gradlew_path = repository_root / "gradlew"
        self.shadow_jar_directory = repository_root / "build" / "libs"

    @property
    def compatibility_response(self) -> str:
        return self.COMPATIBILITY_RESPONSE

    def build_shadow_jar(self) -> Path:
        result = self.runner.run(
            [str(self.gradlew_path), "--no-daemon", ":dmtools-core:shadowJar"],
            cwd=self.repository_root,
        )
        if result.returncode != 0:
            raise AssertionError(
                "Failed to build the DMTools fat JAR.\n"
                f"stdout:\n{result.stdout}\n\nstderr:\n{result.stderr}"
            )
        return self.find_shadow_jar()

    def find_shadow_jar(self) -> Path:
        jars = sorted(
            self.shadow_jar_directory.glob("*-all.jar"),
            key=lambda candidate: candidate.stat().st_mtime,
            reverse=True,
        )
        if not jars:
            raise FileNotFoundError(
                f"No fat JAR was found under {self.shadow_jar_directory.as_posix()}."
            )
        return jars[0]

    def run_job(
        self,
        job_name: str,
        params: dict[str, Any] | None = None,
    ) -> ProcessExecutionResult:
        jar_path = self.find_shadow_jar() if any(self.shadow_jar_directory.glob("*-all.jar")) else self.build_shadow_jar()
        payload = {
            "name": job_name,
            "params": params or {},
        }

        with tempfile.TemporaryDirectory(prefix="dmtools-job-") as temp_dir:
            config_path = Path(temp_dir) / f"{job_name.lower()}.json"
            config_path.write_text(json.dumps(payload), encoding="utf-8")
            return self.runner.run(
                [
                    "java",
                    "-Dlog4j2.configurationFile=classpath:log4j2-cli.xml",
                    "-Dlog4j.configuration=log4j2-cli.xml",
                    "-Dlog4j2.disable.jmx=true",
                    "-Djava.net.preferIPv4Stack=true",
                    "-Djava.rmi.server.hostname=127.0.0.1",
                    "--add-opens",
                    "java.base/java.lang=ALL-UNNAMED",
                    "-XX:-PrintWarnings",
                    "-Dpolyglot.engine.WarnInterpreterOnly=false",
                    "-cp",
                    str(jar_path),
                    "com.github.istin.dmtools.job.JobRunner",
                    "run",
                    str(config_path),
                ],
                cwd=self.repository_root,
                trace_network=True,
            )

    def parse_result(self, execution: ProcessExecutionResult) -> list[dict[str, Any]]:
        return json.loads(execution.stdout)

    def outbound_network_lines(
        self,
        execution: ProcessExecutionResult,
    ) -> list[str]:
        return [
            line
            for line in execution.trace_lines
            if ("AF_INET" in line or "AF_INET6" in line)
            and self._NETWORK_ACTIVITY_PATTERN.search(line)
        ]
